### PR TITLE
An author by another name

### LINKED ISSUE #1213: cargo new does not detect author name and email for author field
```
/tmp$ cargo new --bin hello
/tmp$ cat hello/Cargo.toml 
[package]

name = "hello"
version = "0.0.1"
authors = ["josh"]
```

I have `$NAME`, `$GIT_AUTHOR_NAME`, and `$GIT_COMMITTER_NAME` all set, and likewise for EMAIL.  git knows who I am:

```
/tmp/hello$ git add .
/tmp/hello$ git commit -m 'Initial commit'
[master (root-commit) 51b94a2] Initial commit
 3 files changed, 9 insertions(+)
 create mode 100644 .gitignore
 create mode 100644 Cargo.toml
 create mode 100644 src/main.rs
/tmp/hello$ git log
commit 51b94a2fab5045e4c822684f5450f8ede817c2ad (HEAD, master)
Author: Josh Triplett <josh@joshtriplett.org>
Date:   Fri Jan 23 17:12:20 2015 -0800

    Initial commit
```

### PR DESCRIPTION
**Pull request by sbeckeriv**
Dearest Reviewer

This is to support environment variables for name and email more like
git. This closes #1213 . I did look in to using libgit2 but I did not
see a clear way to get the author ident. This just moves from or_else
chaining to map_filter with env checks and getting the nth(0) result.

The docs I found
https://git-scm.com/book/en/v2/Git-Internals-Environment-Variables#Committing
do not really indicate an order of preference to the variables. I am
more then happy to rearrange the arrays in anyway.

I also did not add in a new test for the positive case because the
test for the current variables cover that path.

Thanks for reviewing,
Becker

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks! Maybe this functionality could be refactored a bit as well? (e.g. get an env var from a list of possible keys).

Also, could you add a test for one of the `GIT_*` variables too?

**Comment by sbeckeriv** [comment]
Moved to a function and added test for name and email.

File: src/cargo/ops/cargo_new.rs [hunk 1]
```diff
@@ -435,12 +435,20 @@ mod tests {
     Ok(())
 }
 
+fn get_environment_variable(variables: &Vec<&str>)-> Option<String>{
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_new.rs]: `&Vec<T>` is typically more idiomatically written as `&[T]` in this case (you can avoid the `vec!` as well)

File: src/cargo/ops/cargo_new.rs [hunk 2]
```diff
@@ -435,12 +435,20 @@ mod tests {
...
+    variables.iter()
+    .filter_map(|var| env::var(var).ok())
+    .nth(0)
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_new.rs]: For now Cargo aligns on the `.` for chains like this, and perhaps `.nth(0)` could get replaced with `next()`?

File: src/cargo/ops/cargo_new.rs [hunk 3]
```diff
@@ -449,8 +457,9 @@ fn discover_author() -> CargoResult<(String, Option<String>)> {
                   username_var)
         }
     };
+    let email_variables = vec!["EMAIL", "GIT_AUTHOR_EMAIL", "GIT_COMMITTER_EMAIL"];
     let email = git_config.and_then(|g| g.get_string("user.email").ok())
-                          .or_else(|| env::var("EMAIL").ok());
+                         .or_else(|| get_environment_variable(&email_variables) );
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_new.rs]: Stray space removal?

File: src/cargo/ops/cargo_new.rs (hunk 1 above)
Comment (alexcrichton) [inline src/cargo/ops/cargo_new.rs]: Could you add a space after this `>` as well?

**Comment by sbeckeriv** [comment]
Thanks for the comments. They are all addressed. I also was trying to figure out how to write the array signature. Thanks for that! 

Becker

**Comment by joshtriplett** [comment]
Thanks for working on this!

@sbeckeriv wrote:
> The docs I found
https://git-scm.com/book/en/v2/Git-Internals-Environment-Variables#Committing
do not really indicate an order of preference to the variables.

"man git-commit-tree" provides the full precedence:

```
       author and
       committer information is taken from the following environment
       variables, if set:

           GIT_AUTHOR_NAME
           GIT_AUTHOR_EMAIL
           GIT_AUTHOR_DATE
           GIT_COMMITTER_NAME
           GIT_COMMITTER_EMAIL
           GIT_COMMITTER_DATE

       (nb "<", ">" and "\n"s are stripped)

       In case (some of) these environment variables are not set, the
       information is taken from the configuration items user.name and
       user.email, or, if not present, the environment variable EMAIL, or, if
       that is not set, system user name and the hostname used for outgoing
       mail (taken from /etc/mailname and falling back to the fully qualified
       hostname when that file does not exist).
```

I don't think the last fallback (to the username, mailname, and FQDN) makes sense anymore.  So the git precedence for name is `$GIT_AUTHOR_NAME` or `$GIT_COMMITTER_NAME`, then `user.name`; the git precedence for email is `$GIT_AUTHOR_EMAIL` or `$GIT_COMMITTER_EMAIL`, then `user.email`, then `$EMAIL`.

Adding a name fallback to the generic `$NAME` makes sense, but it should come after the git-specific environment variables and then `user.name`.  Using `$USER` or `$USERNAME` doesn't make sense, because those only contain a login name, which doesn't make sense in the author field.  As for `AUTHOR` versus `COMMITTER`, I'd tend to favor the former as you did, but it shouldn't matter.  Finally, you might want to add a `$CARGO_NAME` and `$CARGO_EMAIL`, on the off chance someone wants to customize this.  (For instance, someone might use a specific email address for their Rust work.)

So, I'd suggest the following:

Name: `$CARGO_NAME`, `$GIT_AUTHOR_NAME`, `$GIT_COMMITTER_NAME`, `user.name`, `$NAME`

Email: `$CARGO_EMAIL`, `$GIT_AUTHOR_EMAIL`, `$GIT_COMMITTER_EMAIL`, `user.email`, `$EMAIL`.

In both cases, the presence of the config item in the middle of the list makes it necessary to split the array, or introduce an enum, or go back to `and_then`.

**Pull request merged**